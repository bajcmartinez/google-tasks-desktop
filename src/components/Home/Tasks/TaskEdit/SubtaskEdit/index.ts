import SubtaskEdit from './SubtaskEdit';
export default SubtaskEdit;