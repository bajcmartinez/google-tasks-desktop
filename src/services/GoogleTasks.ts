import moment, { Moment } from 'moment';

// @ts-ignore
let google = window.gapi;

export type TaskList = {
  id: string,
  title: string,
  updatedAt: Moment,
  status: string
}

export type Task = {
  id: string,
  title: string,
  notes?: string,
  completed: boolean,
  completedAt?: Moment,
  dueAt?: Moment,
  parent: string,
  updatedAt: Moment,
  status: string,
  listId: string,
  subtasks: Task[]
}

class GoogleTasksService {
  private clientId: string = "721709625729-0jp536rce8pn3i5ie0pg213d2t55mu55.apps.googleusercontent.com";
  private scopes: string = 'https://www.googleapis.com/auth/tasks';
  private isLoaded: boolean = false;
  private auth: any;

  /**
   * Loads the client library and gets all the api required information from google servers
   *
   */
  load() {
    if (this.isLoaded) return Promise.resolve()

    const self = this;

    return new Promise(resolve => {

      // To load first we need to inject the scripts
      const script = document.createElement("script");
      script.src = "gapi.js";
      // @ts-ignore
      script.onload = () => {
        // @ts-ignore
        google = window.gapi;

        // Then we load the API
        google.load('client', () => {
          google.client.load('tasks', 'v1', () => {
            // @ts-ignore
            google = window.gapi;

            self.isLoaded = true;
            resolve();
          });
        });
      };

      document.body.appendChild(script);
    });
  }

  /**
   * Gets the client authorization to query google's API
   *
   */
  authorize() {
    return new Promise(async (resolve) => {
      await this.load();

      this.auth = await google.auth2.init({
        client_id: this.clientId,
        scope: this.scopes
      });

      resolve();
    });
  }

  /**
   * Returns whether the current session is signed in or not
   *
   */
  isSignedIn () {
    if (!this.auth) return false;
    return this.auth.isSignedIn.get();
  }

  /**
   * Event listener for sign in status
   *
   * @param subscriber
   */
  subscribeSigninStatus (subscriber: (status: boolean) => void) {
    if (!this.auth) return false;
    return this.auth.isSignedIn.listen(subscriber);
  }

  /**
   * Starts the sign in process against your Google Account
   *
   */
  signIn() {
    this.auth.signIn();
  }

  /**
   * Lists all tasks lists
   *
   * @returns TaskList[]
   */
  async listTaskLists() {
    const response = await google.client.tasks.tasklists.list();

    return response.result.items.map((item: any): TaskList => ({
      id: item["id"],
      title: item["title"],
      updatedAt: moment(item["updated"])
    }) as TaskList);
  }

  /**
   * Lists all tasks for a given task list
   *
   * @returns Task[]
   */
  async listTasks(taskListId: string) {
    const response = await google.client.tasks.tasks.list({
      tasklist: taskListId,
      showCompleted: true,
      showHidden: true
    });

    const items = response.result.items;

    const mapItems = (tasks: any[]): Task[] => {
      return tasks.map((item: any): Task => ({
        id: item["id"],
        title: item["title"],
        notes: item["notes"],
        dueAt: item["due"] ? moment(item["due"]) : undefined,
        parent: item["parent"],
        completed: item["status"] === "completed",
        completedAt: item["completed"] ? moment(item["completed"]) : undefined,
        updatedAt: moment(item["updated"]),
        listId: taskListId,
        status: item["status"],
        subtasks: mapItems(items.filter((subitem: any) => subitem["parent"] === item["id"]))
      }) as Task);
    };

    return mapItems(items);
  }

  async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    await google.client.tasks.tasks.update({
      tasklist,
      task,
      id: task,
      status: completed ? 'completed' : 'needsAction'
    });
  }
}

export default new GoogleTasksService();