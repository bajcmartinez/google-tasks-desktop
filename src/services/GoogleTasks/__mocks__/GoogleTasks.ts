class GoogleTasksService {

  isSignedIn() {
    return Promise.resolve(true);
  }

  listTaskLists() {
    return Promise.resolve([]);
  }
}

export default new GoogleTasksService();