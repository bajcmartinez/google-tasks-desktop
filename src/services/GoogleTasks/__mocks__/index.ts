import moment from 'moment';
import { Task } from "../../../types/google";
import { GoogleTasksWebService } from '../Web';

class Index extends GoogleTasksWebService {

  private _signedIn: boolean = false;
  private _subscription: ((status: boolean) => void) | undefined;

  load(callback: (isSignedIn: boolean) => void) {
    this.subscribeSigninStatus(callback);
    return Promise.resolve();
  }

  authorize() {
    return Promise.resolve(true);
  }

  signIn() {
    this._signedIn = true;
    this._subscription && this._subscription(this._signedIn);
  }

  signOut() {
    this._signedIn = false;
    this._subscription && this._subscription(this._signedIn);
  }

  isSignedIn() {
    return this._signedIn;
  }

  listTaskLists() {
    return Promise.resolve([
      {
        id: "1",
        status: "active",
        title: "Task List 1",
        updatedAt: moment()
      },
      {
        id: "2",
        status: "active",
        title: "Task List 2",
        updatedAt: moment()
      }
    ]);
  }

  async listTasks(taskListId: string, pageToken: string = '') {
    return Promise.resolve([
      {
        id: `${taskListId.toString()}_1`,
        title: 'Task 1',
        notes: 'notes 1',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(7, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_2`,
        title: 'Task 2',
        notes: 'notes 2',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_3`,
        title: 'Task 3',
        notes: 'notes 3',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(-3, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_4`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_5`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "completed",
        updatedAt: moment(),
        parent: `${taskListId.toString()}_4`
      }
    ]);
  }

  async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    return Promise.resolve();
  }

  async insertTask(task: Task) {
    return Promise.resolve({
      result: {
        id: 999
      }
    });
  }

  async deleteTask(task: string, tasklist: string) {
    return Promise.resolve();
  }

  async updateTask(task: Task) {
    return Promise.resolve();
  }

  subscribeSigninStatus (subscriber: (status: boolean) => void) {
    this._subscription = subscriber;
  }

  reset() {
    this._signedIn = false;
    // @ts-ignore
    this._subscription = null;
  }
}

const r = new Index();
export default r;