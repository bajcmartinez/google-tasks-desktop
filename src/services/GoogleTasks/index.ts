/* istanbul ignore file */
import { GoogleTasksWebService } from './Web'

let GoogleTasksService = GoogleTasksWebService;

if (typeof process !== 'undefined' && process.versions && process.versions.hasOwnProperty('electron')) {
  console.log('here I am');
  const { GoogleTasksElectronService } = require('./Electron');
  GoogleTasksService = GoogleTasksElectronService;
}

const r = new GoogleTasksService();
export default r;