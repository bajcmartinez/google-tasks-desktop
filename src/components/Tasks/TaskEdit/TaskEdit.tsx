import React, { ChangeEvent, useEffect, useState } from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import CalendarIcon from '@material-ui/icons/Event';
import TextField from '@material-ui/core/TextField';
import { Grid } from '@material-ui/core';
import MenuItem from '@material-ui/core/MenuItem';
import { KeyboardDatePicker } from '@material-ui/pickers';
import { Moment } from 'moment';
import { debounce } from 'throttle-debounce';
import Typography from '@material-ui/core/Typography';
import List from '@material-ui/core/List';
import Button from '@material-ui/core/Button';
import DeleteIcon from '@material-ui/icons/Delete';
import AddIcon from '@material-ui/icons/Add';
import IconButton from '@material-ui/core/IconButton';
import SubtaskEdit from './SubtaskEdit';
import { Task, TaskList } from '../../../types/google';

interface IProps {
  task?: Task;
  taskLists: TaskList[];
  insertTask: (task: Task) => void;
  updateTask: (task: Task) => void;
  deleteTask: (task: Task) => void;
}

const useStyles = makeStyles(theme => ({
  multiline: {
    minHeight: 60,
  },

  due: {
    ...theme.typography.body2,
    color: theme.palette.text.secondary,
    textTransform: 'none',
    marginTop: theme.spacing(1),
  },

  past: {
    color: '#d93025',
    marginRight: theme.spacing(1),
  },

  future: {
    color: '#1a73e8',
    marginRight: theme.spacing(1),
  },

  button: {
    margin: theme.spacing(1),
  },

  rightIcon: {
    marginLeft: theme.spacing(1),
  },

  subtasks: {
    display: 'flex',
  },

  spacer: {
    flexGrow: 1,
  },
}));

const TaskEdit: React.FC<IProps> = ({ task, taskLists, insertTask, updateTask, deleteTask }) => {
  const classes = useStyles();

  const [form, setForm] = useState<Task | null>(null);
  const [debounced, setDebounced] = useState();

  // Rescope variables for debounce
  useEffect(() => {
    setDebounced(() =>
      debounce(500, (updatedTask: Task | null) => {
        if (updatedTask && updatedTask.isDirty) {
          updateTask({
            ...updatedTask,
            isDirty: false,
          });
        }
      }),
    );
  }, [updateTask]);

  let titleInput: HTMLInputElement | null = null;

  // Reset the form when the task props changes
  useEffect(() => {
    // Update the edit form only if we have selected a new task
    // This prevents the screen from updating and replacing the user content with the background update
    if (!form || !task || form.id !== task.id) {
      setForm(task ? { ...task } : null);
      if (titleInput) titleInput.focus();
    }
    if (form && task && form.subtasks.length !== task.subtasks.length) {
      setForm({
        ...form,
        subtasks: task.subtasks,
      });
    }
  }, [task, form, titleInput]);

  if (!form) {
    return <div>No task selected</div>;
  }

  const handleInsertTask = (): void => {
    insertTask({
      title: '',
      notes: '',
      parent: form.id,
      listId: form.listId,
      subtasks: [] as Task[],
    } as Task);
  };

  const handleChange = (name: string) => (event: ChangeEvent<HTMLInputElement>) => {
    const newTask = {
      ...form,
      [name]: event.target.value,
      isDirty: true,
    };

    setForm(newTask);
    debounced(newTask);
  };

  const handleDueDateChange = (date: Moment | null) => {
    const newTask = {
      ...form,
      dueAt: date || undefined,
      isDirty: true,
    };
    setForm(newTask);
    debounced(newTask);
  };

  const due = form.dueAt ? (
    <CalendarIcon className={form.dueAt.isBefore() ? classes.past : classes.future} />
  ) : (
    <CalendarIcon />
  );

  return (
    <form>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <TextField
            inputRef={(input: HTMLInputElement) => {
              titleInput = input;
            }}
            label="Title"
            value={form.title}
            data-test-id="task-edit-title"
            onChange={handleChange('title')}
            fullWidth
          />
        </Grid>

        <Grid item xs={12}>
          <TextField
            variant="outlined"
            label="Details"
            value={form.notes}
            multiline
            inputProps={{ className: classes.multiline }}
            onChange={handleChange('notes')}
            fullWidth
          />
        </Grid>

        <Grid item xs={12}>
          <TextField label="List" value={form.listId} onChange={handleChange('listId')} select fullWidth>
            {taskLists.map(taskList => (
              <MenuItem key={taskList.id} value={taskList.id}>
                {taskList.title}
              </MenuItem>
            ))}
          </TextField>
        </Grid>

        <Grid item xs={12}>
          <KeyboardDatePicker
            variant="inline"
            label="Due At"
            value={form.dueAt ? form.dueAt : null}
            format="MMM DD, YYYY"
            onChange={handleDueDateChange}
            keyboardIcon={due}
          />
        </Grid>

        <Grid item xs={12}>
          <div className={classes.subtasks}>
            <Typography variant="h5">Subtasks</Typography>
            <div className={classes.spacer} />
            <IconButton title="Add Subtask" aria-label="Add Subtask" size="small" onClick={handleInsertTask}>
              <AddIcon />
            </IconButton>
          </div>

          <List component="ul">
            {form.subtasks.length > 0 ? (
              form.subtasks.map((subtask: Task) => (
                <SubtaskEdit key={subtask.id} updateTask={updateTask} deleteTask={deleteTask} task={subtask} />
              ))
            ) : (
              <Typography variant="body1">No sub-tasks</Typography>
            )}
          </List>
        </Grid>

        <Grid item xs={12}>
          <Button
            variant="contained"
            color="default"
            className={classes.button}
            data-test-id="task-delete-button"
            onClick={() => deleteTask(form)}
          >
            Delete Task
            <DeleteIcon className={classes.rightIcon} />
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};

export default TaskEdit;
