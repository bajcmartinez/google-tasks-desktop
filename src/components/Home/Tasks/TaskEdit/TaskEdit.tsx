import React, {ChangeEvent, useEffect, useRef} from 'react'
import makeStyles from '@material-ui/core/styles/makeStyles';
import { Task, TaskList } from '../../../../services/GoogleTasks';
import CalendarIcon  from '@material-ui/icons/CalendarToday';
import TextField from '@material-ui/core/TextField';
import { Grid } from '@material-ui/core'
import MenuItem from '@material-ui/core/MenuItem'
import { DatePicker, MaterialUiPickersDate } from '@material-ui/pickers'
import { Moment } from 'moment'
import InputAdornment from '@material-ui/core/InputAdornment'
import { debounce } from 'throttle-debounce'

interface IProps {
  task?: Task,
  taskLists: TaskList[],
  updateTask: (task: Task) => void
}

const useStyles = makeStyles(theme => ({
  multiline: {
    minHeight: 60
  },

  due: {
    ...theme.typography.body2,
    color: theme.palette.text.secondary,
    textTransform: 'none',
    marginTop: theme.spacing(1)
  },

  past: {
    color: '#d93025',
    marginRight: theme.spacing(1)
  },

  future: {
    color: '#1a73e8',
    marginRight: theme.spacing(1)
  },

  buttons: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
}));

const TaskEdit: React.FC<IProps> = (props) => {
  const classes = useStyles();

  const [task, setTask] = React.useState<Task | null>(null);
  const debounced = useRef(debounce(500, (task: Task | null) => {
    if (task && task.isDirty) {
      props.updateTask({
        ...task,
        isDirty: false
      });
    }
  }));

  // Reset the form when the task props changes
  useEffect(() => {
    setTask(props.task ? { ...props.task } : null);
  }, [props.task]);

  // Save when the task changes
  useEffect(() => {
    debounced.current(task);
  }, [task]);

  if (!task) {
    return <div>No task selected</div>;
  }

  const handleChange = (name:string) => (event:ChangeEvent<HTMLInputElement>) => {
    setTask({
      ...task,
      [name]: event.target.value,
      isDirty: true
    });
  };

  const handleDueDateChange = (date: MaterialUiPickersDate) => {
    setTask({
      ...task,
      dueAt: date as Moment,
      isDirty: true
    });
  };

  const taskEdit = { ...task } as Task;

  const due = task.dueAt ? (
    <CalendarIcon className={task.dueAt.isBefore() ? classes.past : classes.future} />
    ) : (
    <CalendarIcon />
  );

  return (
    <form>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <TextField
            label="Title"
            value={taskEdit.title}
            onChange={handleChange('title')}
            fullWidth
          />
        </Grid>

        <Grid item xs={12}>
          <TextField
            variant="outlined"
            label="Details"
            value={taskEdit.notes}
            multiline
            inputProps={{className: classes.multiline}}
            onChange={handleChange('notes')}
            fullWidth
          />
        </Grid>

        <Grid item xs={12}>
          <TextField
            label="List"
            value={taskEdit.listId}
            onChange={handleChange('listId')}
            select
            fullWidth
          >
            {props.taskLists.map(taskList => (
              <MenuItem key={taskList.id} value={taskList.id}>
                {taskList.title}
              </MenuItem>
            ))}
          </TextField>
        </Grid>

        <Grid item xs={12}>
          <DatePicker
            clearable
            label="Due At"
            value={taskEdit.dueAt ? taskEdit.dueAt : null}
            onChange={handleDueDateChange}
            InputProps={{
              endAdornment: (
                <InputAdornment position="end">
                  {due}
                </InputAdornment>
              ),
            }}
          />
        </Grid>
      </Grid>
    </form>
  );
}

export default TaskEdit;
