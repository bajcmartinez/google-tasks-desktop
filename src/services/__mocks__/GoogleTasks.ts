import moment from 'moment';
import {Task} from "../GoogleTasks";

class GoogleTasksService {

  static _signedIn: boolean = false;
  static _subscription: (status: boolean) => void | undefined;

  static authorize() {
    return Promise.resolve(true);
  }

  static signIn() {
    this._signedIn = true;
    this._subscription && this._subscription(this._signedIn);
  }

  static signOut() {
    this._signedIn = false;
    this._subscription && this._subscription(this._signedIn);
  }

  static isSignedIn() {
    return this._signedIn;
  }

  static listTaskLists() {
    return Promise.resolve([
      {
        id: "1",
        status: "active",
        title: "Task List 1",
        updatedAt: moment()
      },
      {
        id: "2",
        status: "active",
        title: "Task List 2",
        updatedAt: moment()
      }
    ]);
  }

  static async listTasks(taskListId: string, pageToken: string = '') {
    return Promise.resolve([
      {
        id: `${taskListId.toString()}_1`,
        title: 'Task 1',
        notes: 'notes 1',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(7, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_2`,
        title: 'Task 2',
        notes: 'notes 2',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_3`,
        title: 'Task 3',
        notes: 'notes 3',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(-3, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_4`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "needsAction",
        updatedAt: moment(),
        parent: ""
      },
      {
        id: `${taskListId.toString()}_5`,
        title: 'Task 4',
        notes: 'notes 4',
        subtasks: [],
        listId: taskListId,
        isDirty: false,
        dueAt: moment().add(1, 'days'),
        completed: false,
        status: "completed",
        updatedAt: moment(),
        parent: `${taskListId.toString()}_4`
      }
    ]);
  }

  static async updateTaskCompletion(task: string, tasklist: string, completed: boolean) {
    return Promise.resolve();
  }

  static async insertTask(task: Task) {
    return Promise.resolve({
      result: {
        id: 999
      }
    });
  }

  static async deleteTask(task: string, tasklist: string) {
    return Promise.resolve();
  }

  static async updateTask(task: Task) {
    return Promise.resolve();
  }

  static subscribeSigninStatus (subscriber: (status: boolean) => void) {
    this._subscription = subscriber;
  }

  /**
   * Only used for testing purposes
   */
  static reset() {
    this._signedIn = false;
    // @ts-ignore
    this._subscription = null;
  }
}

export default GoogleTasksService;