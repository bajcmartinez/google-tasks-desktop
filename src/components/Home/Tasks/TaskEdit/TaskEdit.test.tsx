import React, { ChangeEvent } from 'react'
import { shallow, mount } from 'enzyme';
import TaskEdit from './TaskEdit';
import { Task, TaskList } from '../../../../services/GoogleTasks'
import moment from 'moment';
import MomentUtils from '@date-io/moment'
import { MuiPickersUtilsProvider } from '@material-ui/pickers'
import TextField from '@material-ui/core/TextField';
import { act } from 'react-dom/test-utils'
import { Button } from '@material-ui/core'

describe("Basic", () => {
  it('should render without crashing', () => {
    const task: Task = {
      id: "1",
      title: "My Title",
      completed: false,
      parent: "",
      updatedAt: moment(),
      status: 'neededAction',
      listId: '1',
      subtasks: [{
        id: "2",
        title: "My subtask",
        completed: false,
        parent: "",
        updatedAt: moment(),
        status: 'neededAction',
        listId: '1',
        isDirty: false
      } as Task],
      isDirty: false
    }

    const taskEdit = shallow(
      <TaskEdit
        task={task}
        taskLists={[]}
        deleteTask={() => null}
        updateTask={() => null}
      />
    );
    expect(taskEdit.exists()).toBe(true);
  });

  it('should render a <TaskEdit /> component as expected', () => {
    const task: Task = {
      id: "1",
      title: "My Title",
      completed: false,
      parent: "",
      updatedAt: moment(),
      status: 'neededAction',
      listId: '1',
      subtasks: [{
        id: "2",
        title: "My subtask",
        completed: false,
        parent: "",
        updatedAt: moment(),
        status: 'neededAction',
        listId: '1',
        isDirty: false
      } as Task],
      isDirty: false
    }

    const taskLists: TaskList[] = [{
      id: "1",
      status: "active",
      title: "Task List 1",
      updatedAt: moment()
    }]

    const taskEdit = mount(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <TaskEdit
          task={task}
          taskLists={taskLists}
          deleteTask={() => null}
          updateTask={() => null}
        />
      </MuiPickersUtilsProvider>
    );

    expect(taskEdit).toMatchSnapshot();
  });
});

describe("Events", () => {

  it('should trigger the text change event', () => {
    jest.useFakeTimers();

    const handler = jest.fn();

    const task: Task = {
      id: "1",
      title: "My Title",
      completed: false,
      parent: "",
      updatedAt: moment(),
      status: 'neededAction',
      listId: '1',
      subtasks: [],
      isDirty: false
    }

    const taskEdit = mount(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <TaskEdit
          task={task}
          taskLists={[]}
          deleteTask={() => null}
          updateTask={handler}
        />
      </MuiPickersUtilsProvider>
    );

    const textField = taskEdit.find(TextField).first().props();

    act(() => {
      textField.onChange && textField.onChange({
        target: {
          value: "new title"
        }
      } as ChangeEvent<HTMLInputElement>);
    });

    // As events are debounced, let's wait a bit
    jest.runAllTimers();

    expect(handler).toBeCalled();
  });

  it('should trigger the delete event', () => {
    const handler = jest.fn();

    const task: Task = {
      id: "1",
      title: "My Title",
      completed: false,
      parent: "",
      updatedAt: moment(),
      status: 'neededAction',
      listId: '1',
      subtasks: [],
      isDirty: false
    }

    const taskEdit = mount(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <TaskEdit
          task={task}
          taskLists={[]}
          deleteTask={handler}
          updateTask={() => null}
        />
      </MuiPickersUtilsProvider>
    );

    const deleteButton = taskEdit.find(Button).first();
    deleteButton.simulate('click');


    expect(handler).toBeCalled();
  });

});
