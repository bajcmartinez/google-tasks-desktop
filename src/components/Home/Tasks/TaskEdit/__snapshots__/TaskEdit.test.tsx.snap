// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic should render a <TaskEdit /> component as expected 1`] = `ReactWrapper {}`;
